ratis-client/src/main/java/org/apache/ratis/client/impl/RaftClientImpl.java

```diff
@@ -133,6 +133,7 @@ void set(Collection<RaftPeer> newPeers) {
   private final RaftGroupId groupId;
   private final RetryPolicy retryPolicy;
 
+  @SuppressWarnings("java:S3077")
```

Comment:
It seem that we may use @Immutable to avoid it according to https://community.sonarsource.com/t/rule-s3077-frequently-generates-a-false-positive/27479/14
```java
+++ b/ratis-common/src/main/java/org/apache/ratis/protocol/RaftPeerId.java
@@ -22,6 +22,7 @@ import org.apache.ratis.thirdparty.com.google.protobuf.ByteString;
 import org.apache.ratis.util.JavaUtils;
 import org.apache.ratis.util.Preconditions;
 
+import javax.annotation.concurrent.Immutable;
 import java.nio.charset.StandardCharsets;
 import java.util.Map;
 import java.util.Objects;
@@ -33,6 +34,7 @@ import java.util.function.Supplier;
  * <p>
  * This is a value-based class.
  */
+@Immutable
 public final class RaftPeerId {
```

Reply:
Good idea!